Initial Prometheus Metrics Endpoint

jjcollinge opened the pull request:
# Description

I've added an initial metrics endpoint using OpenCensus's Prometheus exporter to hook into Dapr's HTTP pipeline. Due to the incompatibility between fasthttp, net/http, and ochttp this is a bit more messy than I'd like. I'll continue to investigate better alternatives.

This is built on the existing work in #950 

Example `/metrics` endpoint:
```
# HELP nodeapp_opencensus_io_http_server_latency Latency distribution of HTTP requests
# TYPE nodeapp_opencensus_io_http_server_latency histogram
nodeapp_opencensus_io_http_server_latency_bucket{le="1"} 2
nodeapp_opencensus_io_http_server_latency_bucket{le="2"} 5
nodeapp_opencensus_io_http_server_latency_bucket{le="3"} 9
nodeapp_opencensus_io_http_server_latency_bucket{le="4"} 12
nodeapp_opencensus_io_http_server_latency_bucket{le="5"} 13
nodeapp_opencensus_io_http_server_latency_bucket{le="6"} 14
nodeapp_opencensus_io_http_server_latency_bucket{le="8"} 14
nodeapp_opencensus_io_http_server_latency_bucket{le="10"} 14
nodeapp_opencensus_io_http_server_latency_bucket{le="13"} 14
nodeapp_opencensus_io_http_server_latency_bucket{le="16"} 14
nodeapp_opencensus_io_http_server_latency_bucket{le="20"} 15
nodeapp_opencensus_io_http_server_latency_bucket{le="25"} 15
nodeapp_opencensus_io_http_server_latency_bucket{le="30"} 15
nodeapp_opencensus_io_http_server_latency_bucket{le="40"} 15
nodeapp_opencensus_io_http_server_latency_bucket{le="50"} 15
nodeapp_opencensus_io_http_server_latency_bucket{le="65"} 15
nodeapp_opencensus_io_http_server_latency_bucket{le="80"} 15
nodeapp_opencensus_io_http_server_latency_bucket{le="100"} 15
nodeapp_opencensus_io_http_server_latency_bucket{le="130"} 15
nodeapp_opencensus_io_http_server_latency_bucket{le="160"} 15
nodeapp_opencensus_io_http_server_latency_bucket{le="200"} 15
nodeapp_opencensus_io_http_server_latency_bucket{le="250"} 15
nodeapp_opencensus_io_http_server_latency_bucket{le="300"} 15
nodeapp_opencensus_io_http_server_latency_bucket{le="400"} 15
nodeapp_opencensus_io_http_server_latency_bucket{le="500"} 15
nodeapp_opencensus_io_http_server_latency_bucket{le="650"} 15
nodeapp_opencensus_io_http_server_latency_bucket{le="800"} 15
nodeapp_opencensus_io_http_server_latency_bucket{le="1000"} 15
nodeapp_opencensus_io_http_server_latency_bucket{le="2000"} 15
nodeapp_opencensus_io_http_server_latency_bucket{le="5000"} 15
nodeapp_opencensus_io_http_server_latency_bucket{le="10000"} 15
nodeapp_opencensus_io_http_server_latency_bucket{le="20000"} 15
nodeapp_opencensus_io_http_server_latency_bucket{le="50000"} 15
nodeapp_opencensus_io_http_server_latency_bucket{le="100000"} 15
nodeapp_opencensus_io_http_server_latency_bucket{le="+Inf"} 15
nodeapp_opencensus_io_http_server_latency_sum 53.108200000000004
nodeapp_opencensus_io_http_server_latency_count 15
# HELP nodeapp_opencensus_io_http_server_request_bytes Size distribution of HTTP request body
# TYPE nodeapp_opencensus_io_http_server_request_bytes histogram
nodeapp_opencensus_io_http_server_request_bytes_bucket{le="1024"} 15
nodeapp_opencensus_io_http_server_request_bytes_bucket{le="2048"} 15
nodeapp_opencensus_io_http_server_request_bytes_bucket{le="4096"} 15
nodeapp_opencensus_io_http_server_request_bytes_bucket{le="16384"} 15
nodeapp_opencensus_io_http_server_request_bytes_bucket{le="65536"} 15
nodeapp_opencensus_io_http_server_request_bytes_bucket{le="262144"} 15
nodeapp_opencensus_io_http_server_request_bytes_bucket{le="1.048576e+06"} 15
nodeapp_opencensus_io_http_server_request_bytes_bucket{le="4.194304e+06"} 15
nodeapp_opencensus_io_http_server_request_bytes_bucket{le="1.6777216e+07"} 15
nodeapp_opencensus_io_http_server_request_bytes_bucket{le="6.7108864e+07"} 15
nodeapp_opencensus_io_http_server_request_bytes_bucket{le="2.68435456e+08"} 15
nodeapp_opencensus_io_http_server_request_bytes_bucket{le="1.073741824e+09"} 15
nodeapp_opencensus_io_http_server_request_bytes_bucket{le="4.294967296e+09"} 15
nodeapp_opencensus_io_http_server_request_bytes_bucket{le="+Inf"} 15
nodeapp_opencensus_io_http_server_request_bytes_sum 0
nodeapp_opencensus_io_http_server_request_bytes_count 15
# HELP nodeapp_opencensus_io_http_server_request_count Count of HTTP requests started
# TYPE nodeapp_opencensus_io_http_server_request_count counter
nodeapp_opencensus_io_http_server_request_count 16
# HELP nodeapp_opencensus_io_http_server_request_count_by_method Server request count by HTTP method
# TYPE nodeapp_opencensus_io_http_server_request_count_by_method counter
nodeapp_opencensus_io_http_server_request_count_by_method{http_method="GET"} 16
# HELP nodeapp_opencensus_io_http_server_response_bytes Size distribution of HTTP response body
# TYPE nodeapp_opencensus_io_http_server_response_bytes histogram
nodeapp_opencensus_io_http_server_response_bytes_bucket{le="1024"} 15
nodeapp_opencensus_io_http_server_response_bytes_bucket{le="2048"} 15
nodeapp_opencensus_io_http_server_response_bytes_bucket{le="4096"} 15
nodeapp_opencensus_io_http_server_response_bytes_bucket{le="16384"} 15
nodeapp_opencensus_io_http_server_response_bytes_bucket{le="65536"} 15
nodeapp_opencensus_io_http_server_response_bytes_bucket{le="262144"} 15
nodeapp_opencensus_io_http_server_response_bytes_bucket{le="1.048576e+06"} 15
nodeapp_opencensus_io_http_server_response_bytes_bucket{le="4.194304e+06"} 15
nodeapp_opencensus_io_http_server_response_bytes_bucket{le="1.6777216e+07"} 15
nodeapp_opencensus_io_http_server_response_bytes_bucket{le="6.7108864e+07"} 15
nodeapp_opencensus_io_http_server_response_bytes_bucket{le="2.68435456e+08"} 15
nodeapp_opencensus_io_http_server_response_bytes_bucket{le="1.073741824e+09"} 15
nodeapp_opencensus_io_http_server_response_bytes_bucket{le="4.294967296e+09"} 15
nodeapp_opencensus_io_http_server_response_bytes_bucket{le="+Inf"} 15
nodeapp_opencensus_io_http_server_response_bytes_sum 7207
nodeapp_opencensus_io_http_server_response_bytes_count 15
# HELP nodeapp_opencensus_io_http_server_response_count_by_status_code Server response count by status code
# TYPE nodeapp_opencensus_io_http_server_response_count_by_status_code counter
nodeapp_opencensus_io_http_server_response_count_by_status_code{http_status="200"} 15
```

## Issue reference

Please reference the issue this PR will close: #959 

## Checklist

Please make sure you've  completed the relevant tasks for this PR, out of the following list:

* [x] Code compiles correctly
* [ ] Created/updated tests
* [x] Unit tests passing
* [ ] End-to-end tests passing
* [ ] Extended the documentation
* [ ] Specification has been updated
* [ ] Provided sample for the feature

youngbupark commented:
@jjcollinge @yaron2

I reviewed this design. `pull-mode` design may not be right for dapr sidecar.

Let's start with the user scenario.

`As a devops, I can aggregate metrics from dapr runtime sidecars and see Dashboard on Grafana.`

Dapr runtime is a sidecar, which means user can deploy hundreds or thousands of pods to their cluster.

Let's say that I deploy 100 pods to my k8s cluster and I want to aggregate all metrics from 100 pods and set up grafana to see these metrics. pod/service name can be dynamically changed. we cannot expect the name. Then how can we configure `pull mode` for 100 pods in promethus ? 

Prometheus supports both `pull-model` and `push model` (opencensus supports both accordingly). Pull-model for operator, injector, placement, and sentry does make sense, but not dapr runtime sidecar.  Adopting push-mode in dapr side car is the right design rather than pull-mode.

How do you think ?

yaron2 commented:
> @jjcollinge @yaron2
> 
> I reviewed this design. `pull-mode` design may not be right for dapr sidecar.
> 
> Let's start with the user scenario.
> 
> `As a devops, I can aggregate metrics from dapr runtime sidecars and see Dashboard on Grafana.`
> 
> Dapr runtime is a sidecar, which means users can deploy hundreds or thousands of pods to their cluster.
> 
> Let's say that I deploy 100 pods to my k8s cluster and I want to aggregate all metrics from 100 pods and set up grafana to see these metrics. pod/service name can be dynamically changed. we cannot expect the name. Then how can we configure `pull mode` for 100 pods in promethus ?
> 
> Prometheus supports both `pull-model` and `push model` (opencensus supports both accordingly). Pull-model for operator, injector, placement, and sentry does make sense, but not dapr runtime sidecar. Adopting push-mode in dapr side car is the right design rather than pull-mode.
> 
> How do you think ?

I don't see how this is different than any other deployment model with Prometheus in Kubernetes. When you do an aggregate query, you aggregate based on the Deployment ID and/or Dapr ID, which would encompass all the Pods under that deployment.

youngbupark commented:
> > @jjcollinge @yaron2
> > I reviewed this design. `pull-mode` design may not be right for dapr sidecar.
> > Let's start with the user scenario.
> > `As a devops, I can aggregate metrics from dapr runtime sidecars and see Dashboard on Grafana.`
> > Dapr runtime is a sidecar, which means users can deploy hundreds or thousands of pods to their cluster.
> > Let's say that I deploy 100 pods to my k8s cluster and I want to aggregate all metrics from 100 pods and set up grafana to see these metrics. pod/service name can be dynamically changed. we cannot expect the name. Then how can we configure `pull mode` for 100 pods in promethus ?
> > Prometheus supports both `pull-model` and `push model` (opencensus supports both accordingly). Pull-model for operator, injector, placement, and sentry does make sense, but not dapr runtime sidecar. Adopting push-mode in dapr side car is the right design rather than pull-mode.
> > How do you think ?
> 
> I don't see how this is different than any other deployment model with Prometheus in Kubernetes. When you do an aggregate query, you aggregate based on the Deployment ID and/or Dapr ID, which would encompass all the Pods under that deployment.

@jjcollinge Nevermind. I synced up with @yaron2

youngbupark commented:
@jjcollinge @yaron2 I have never used pull-mode for telemetry backend in the previous products. Good to know. :)

jjcollinge commented:
@youngbupark if we standardize on a Prometheus endpoint (pull) then people can use whatever compatible agents they want to scrape it. Or would you prefer to support other [OC stats exporters](https://opencensus.io/exporters/supported-exporters/go/) which would require both pull and push?

If we just do Prometheus (no other exporters) - do you think it's reasonable to define what metrics you want to expose in the dapr config similar to how the `TracingSpec` works, or would you want to define a component? To me it feels like if we're only supporting Prometheus there is no "building block" and we should just have some metrics toggles in the dapr config such as:
```go
type MetricsSpec struct {
	Enabled                 bool `json:"enabled"`
	Route                   string `json:"route"` // defaults to /metrics
        Namespace               string `json:"namespace"` // defaults to daprID
        Labels                  map string[string]
	RequestCount            bool `json:"requestCount"`
	RequestBytes            bool `json:"requestBytes"`
	ResponseBytes           bool `json:"responseBytes"`
	Latency                 bool `json:"latency"`
	CountByMethod           bool `json:"countByMethod"`
	CountByStatusCode       bool `json:"countByStatusCode"`
}
```
These are just the default ones already supported but as we add others we can optionally add those to the view.

youngbupark commented:
> @youngbupark if we standardize on a Prometheus endpoint (pull) then people can use whatever compatible agents they want to scrape it. Or would you prefer to support other [OC stats exporters](https://opencensus.io/exporters/supported-exporters/go/) which would require both pull and push? This would then require a new "MetricsExporter" type I guess.

I think we need to support both now at the end. I agree that at the first iteration, Prometheus can be the default metric backends in Dapr. In the long term, we need to allow user to use the other oc exporter. 
Some teams or companies have their internal telemetry backend/instrumentation, but doesn't support prometheus pull model. Then they need to develop scraping agent to premetheus metric directly (like AKS oms agent) or a loader from prometheus to their own telemetry backend(stackdriver-prometheus sidecar). Allowing oc exporter will be more flexible for them. We need to discuss it more. 

> If we just do Prometheus (no other exporters) - do you think it's reasonable to define what metrics you want to expose in the dapr config similar to how the `TracingSpec` works, or would you want to define a component? To me it feels like if we're only supporting Prometheus there is no "building block" and we should just have some metrics toggles in the dapr config such as:
> 
> ```go
> type MetricsSpec struct {
> 	Enabled                 bool `json:"enabled"`
> 	Route                   string `json:"route"` // defaults to /metrics
>         Namespace               string `json:"namespace"` // defaults to daprID
>         Labels                  map string[string]
> 	RequestCount            bool `json:"requestCount"`
> 	RequestBytes            bool `json:"requestBytes"`
> 	ResponseBytes           bool `json:"responseBytes"`
> 	Latency                 bool `json:"latency"`
> 	CountByMethod           bool `json:"countByMethod"`
> 	CountByStatusCode       bool `json:"countByStatusCode"`
> }
> ```
> 
> These are just the default ones already supported but as we add others we can optionally add those to the view.

I also prefer defining new CRD config `MetricSpec` similar to `TracingSpec`. but the spec's scheme would be different from what you suggest. Instead of adding `bool` flag in struct, I would group the metrics and add string:string map to be more flexible. 

```
EnabledMetricGroups map string[string]
```

jjcollinge commented:
"I would group the metrics and add string:string map to be more flexible." - could you expand on this a little? Let's say we create a group and call it "server_http" which includes the metrics:
- server_latency
- server_request_count
- server_request_bytes
- server_response_bytes
- server_count_by_method
- server_count_by_status_code

What would we store in the map? Feels like an array of group names is more suitable? Also, wanting to make this as easy to pick up and use as possible - it might be good to use a property such as `useDefaultGroups` which overrides the `EnabledMetricsGroups` and just turns on some predefined defaults. What do you think to something like?

```go
type MetricsSpec struct {
	Enabled              bool              `json:"enabled" yaml:"enabled"`
	Route                bool              `json:"route" yaml:"route"`
	Namespace            bool              `json:"namespace" yaml:"namespace"`
	Labels               map[string]string `json:"labels" yaml:"labels"`
	UseDefaultMetrics    bool              `json:"useDefaultMetrics" yaml:"useDefaultMetrics"`
	EnabledMetricsGroups []string          `json:"enabledMetricsGroups" yaml:"enabledMetricsGroups"`
}
```

youngbupark commented:
> "I would group the metrics and add string:string map to be more flexible." - could you expand on this a little? Let's say we create a group and call it "server_http" which includes the metrics:
> 
> * server_latency
> * server_request_count
> * server_request_bytes
> * server_response_bytes
> * server_count_by_method
> * server_count_by_status_code
> 
> What would we store in the map? Feels like an array of group names is more suitable? Or did I miss something? Also, wanting to make this as easy to pick up and use as possible - it might be good to use a property such as `useDefaultGroups` which overrides the `EnabledMetricsGroups` and just turns on some predefined defaults. What do you think to something like?
> 
> ```go
> type MetricsSpec struct {
> 	Enabled              bool              `json:"enabled" yaml:"enabled"`
> 	Route                bool              `json:"route" yaml:"route"`
> 	Namespace            bool              `json:"namespace" yaml:"namespace"`
> 	Labels               map[string]string `json:"labels" yaml:"labels"`
> 	UseDefaultMetrics    bool              `json:"useDefaultMetrics" yaml:"useDefaultMetrics"`
> 	EnabledMetricsGroups []string          `json:"enabledMetricsGroups" yaml:"enabledMetricsGroups"`
> }
> ```

oh :) Nice catch. Yeah it should be array of string.  I created observability label and several work items. Please review them.
https://github.com/dapr/dapr/labels/area%2Fobservability

Also let's continue to discuss MetricSpec in https://github.com/dapr/dapr/issues/972. We should have a mini-spec and review it with the other contributors.

jjcollinge commented:
will move to `nethttpadapter` from [component-contrib](https://github.com/dapr/components-contrib/pull/193) if that PR gets merged.

jjcollinge commented:
Updated to use current draft metrics spec. This only covers HTTP metrics, gRPC can be added in a follow up PR.

I hooked up a dapr to prometheus and grafana just to check it out too:
![image](https://user-images.githubusercontent.com/7241190/74085718-f4625180-4a73-11ea-9dd5-73d5f45c48f9.png)

jjcollinge commented:
Note to self, I’m not sure this is the right approach and will explore moving to using a dedicated server for the metrics endpoint.

jjcollinge commented:
I've moved to a dedicated server for the metrics as it makes adding metrics middleware more consistent across the gRPC and HTTP piplines. I've also moved away from `ochttp` as that was fairly limited to `http_prometheus` and included gRPC, process and Go metrics too.

TODO: I'd like to find a way to have more control over the labels in both the gRPC and HTTP metrics.

This has required some changes to the `MetricsSpec` so it currently looks like:
```go
type MetricsSpec struct {
	Enabled       bool              `json:"enabled" yaml:"enabled"`
	Path          string            `json:"path" yaml:"path"`
	Namespace     string            `json:"namespace" yaml:"namespace"`
	Port          string            `json:"port" yaml:"port"`
	Labels        map[string]string `json:"labels" yaml:"labels"`
	UseDefaults   bool              `json:"useDefaults" yaml:"useDefaults"`
	MetricsGroups []string          `json:"metricsGroups" yaml:"metricsGroups"`
}
```

```
# HELP go_gc_duration_seconds A summary of the GC invocation durations.
# TYPE go_gc_duration_seconds summary
go_gc_duration_seconds{quantile="0"} 8e-06
go_gc_duration_seconds{quantile="0.25"} 7.4e-05
go_gc_duration_seconds{quantile="0.5"} 0.0001977
go_gc_duration_seconds{quantile="0.75"} 0.0002238
go_gc_duration_seconds{quantile="1"} 0.0002238
go_gc_duration_seconds_sum 0.0005035
go_gc_duration_seconds_count 4
# HELP go_goroutines Number of goroutines that currently exist.
# TYPE go_goroutines gauge
go_goroutines 30
# HELP go_info Information about the Go environment.
# TYPE go_info gauge
go_info{version="go1.13.1"} 1
# HELP go_memstats_alloc_bytes Number of bytes allocated and still in use.
# TYPE go_memstats_alloc_bytes gauge
go_memstats_alloc_bytes 8.244808e+06
# HELP go_memstats_alloc_bytes_total Total number of bytes allocated, even if freed.
# TYPE go_memstats_alloc_bytes_total counter
go_memstats_alloc_bytes_total 1.2017472e+07
# HELP go_memstats_buck_hash_sys_bytes Number of bytes used by the profiling bucket hash table.
# TYPE go_memstats_buck_hash_sys_bytes gauge
go_memstats_buck_hash_sys_bytes 1.448094e+06
# HELP go_memstats_frees_total Total number of frees.
# TYPE go_memstats_frees_total counter
go_memstats_frees_total 37825
# HELP go_memstats_gc_cpu_fraction The fraction of this program's available CPU time used by the GC since the program started.
# TYPE go_memstats_gc_cpu_fraction gauge
go_memstats_gc_cpu_fraction 5.3649505618710834e-05
# HELP go_memstats_gc_sys_bytes Number of bytes used for garbage collection system metadata.
# TYPE go_memstats_gc_sys_bytes gauge
go_memstats_gc_sys_bytes 2.38592e+06
# HELP go_memstats_heap_alloc_bytes Number of heap bytes allocated and still in use.
# TYPE go_memstats_heap_alloc_bytes gauge
go_memstats_heap_alloc_bytes 8.244808e+06
# HELP go_memstats_heap_idle_bytes Number of heap bytes waiting to be used.
# TYPE go_memstats_heap_idle_bytes gauge
go_memstats_heap_idle_bytes 5.5697408e+07
# HELP go_memstats_heap_inuse_bytes Number of heap bytes that are in use.
# TYPE go_memstats_heap_inuse_bytes gauge
go_memstats_heap_inuse_bytes 1.0330112e+07
# HELP go_memstats_heap_objects Number of allocated objects.
# TYPE go_memstats_heap_objects gauge
go_memstats_heap_objects 23826
# HELP go_memstats_heap_released_bytes Number of heap bytes released to OS.
# TYPE go_memstats_heap_released_bytes gauge
go_memstats_heap_released_bytes 5.4280192e+07
# HELP go_memstats_heap_sys_bytes Number of heap bytes obtained from system.
# TYPE go_memstats_heap_sys_bytes gauge
go_memstats_heap_sys_bytes 6.602752e+07
# HELP go_memstats_last_gc_time_seconds Number of seconds since 1970 of last garbage collection.
# TYPE go_memstats_last_gc_time_seconds gauge
go_memstats_last_gc_time_seconds 1.5812413224803903e+09
# HELP go_memstats_lookups_total Total number of pointer lookups.
# TYPE go_memstats_lookups_total counter
go_memstats_lookups_total 0
# HELP go_memstats_mallocs_total Total number of mallocs.
# TYPE go_memstats_mallocs_total counter
go_memstats_mallocs_total 61651
# HELP go_memstats_mcache_inuse_bytes Number of bytes in use by mcache structures.
# TYPE go_memstats_mcache_inuse_bytes gauge
go_memstats_mcache_inuse_bytes 27776
# HELP go_memstats_mcache_sys_bytes Number of bytes used for mcache structures obtained from system.
# TYPE go_memstats_mcache_sys_bytes gauge
go_memstats_mcache_sys_bytes 32768
# HELP go_memstats_mspan_inuse_bytes Number of bytes in use by mspan structures.
# TYPE go_memstats_mspan_inuse_bytes gauge
go_memstats_mspan_inuse_bytes 88128
# HELP go_memstats_mspan_sys_bytes Number of bytes used for mspan structures obtained from system.
# TYPE go_memstats_mspan_sys_bytes gauge
go_memstats_mspan_sys_bytes 98304
# HELP go_memstats_next_gc_bytes Number of heap bytes when next garbage collection will take place.
# TYPE go_memstats_next_gc_bytes gauge
go_memstats_next_gc_bytes 1.6132608e+07
# HELP go_memstats_other_sys_bytes Number of bytes used for other system allocations.
# TYPE go_memstats_other_sys_bytes gauge
go_memstats_other_sys_bytes 1.47465e+06
# HELP go_memstats_stack_inuse_bytes Number of bytes in use by the stack allocator.
# TYPE go_memstats_stack_inuse_bytes gauge
go_memstats_stack_inuse_bytes 1.081344e+06
# HELP go_memstats_stack_sys_bytes Number of bytes obtained from system for stack allocator.
# TYPE go_memstats_stack_sys_bytes gauge
go_memstats_stack_sys_bytes 1.081344e+06
# HELP go_memstats_sys_bytes Number of bytes obtained from system.
# TYPE go_memstats_sys_bytes gauge
go_memstats_sys_bytes 7.25486e+07
# HELP go_threads Number of OS threads created.
# TYPE go_threads gauge
go_threads 20
# HELP grpc_server_handled_total Total number of RPCs completed on the server, regardless of success or failure.
# TYPE grpc_server_handled_total counter
grpc_server_handled_total{grpc_code="Aborted",grpc_method="CallActor",grpc_service="daprinternal.DaprInternal",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="Aborted",grpc_method="CallLocal",grpc_service="daprinternal.DaprInternal",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="Aborted",grpc_method="DeleteState",grpc_service="dapr.Dapr",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="Aborted",grpc_method="GetState",grpc_service="dapr.Dapr",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="Aborted",grpc_method="InvokeBinding",grpc_service="dapr.Dapr",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="Aborted",grpc_method="InvokeService",grpc_service="dapr.Dapr",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="Aborted",grpc_method="PublishEvent",grpc_service="dapr.Dapr",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="Aborted",grpc_method="SaveState",grpc_service="dapr.Dapr",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="Aborted",grpc_method="UpdateComponent",grpc_service="daprinternal.DaprInternal",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="AlreadyExists",grpc_method="CallActor",grpc_service="daprinternal.DaprInternal",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="AlreadyExists",grpc_method="CallLocal",grpc_service="daprinternal.DaprInternal",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="AlreadyExists",grpc_method="DeleteState",grpc_service="dapr.Dapr",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="AlreadyExists",grpc_method="GetState",grpc_service="dapr.Dapr",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="AlreadyExists",grpc_method="InvokeBinding",grpc_service="dapr.Dapr",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="AlreadyExists",grpc_method="InvokeService",grpc_service="dapr.Dapr",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="AlreadyExists",grpc_method="PublishEvent",grpc_service="dapr.Dapr",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="AlreadyExists",grpc_method="SaveState",grpc_service="dapr.Dapr",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="AlreadyExists",grpc_method="UpdateComponent",grpc_service="daprinternal.DaprInternal",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="Canceled",grpc_method="CallActor",grpc_service="daprinternal.DaprInternal",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="Canceled",grpc_method="CallLocal",grpc_service="daprinternal.DaprInternal",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="Canceled",grpc_method="DeleteState",grpc_service="dapr.Dapr",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="Canceled",grpc_method="GetState",grpc_service="dapr.Dapr",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="Canceled",grpc_method="InvokeBinding",grpc_service="dapr.Dapr",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="Canceled",grpc_method="InvokeService",grpc_service="dapr.Dapr",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="Canceled",grpc_method="PublishEvent",grpc_service="dapr.Dapr",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="Canceled",grpc_method="SaveState",grpc_service="dapr.Dapr",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="Canceled",grpc_method="UpdateComponent",grpc_service="daprinternal.DaprInternal",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="DataLoss",grpc_method="CallActor",grpc_service="daprinternal.DaprInternal",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="DataLoss",grpc_method="CallLocal",grpc_service="daprinternal.DaprInternal",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="DataLoss",grpc_method="DeleteState",grpc_service="dapr.Dapr",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="DataLoss",grpc_method="GetState",grpc_service="dapr.Dapr",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="DataLoss",grpc_method="InvokeBinding",grpc_service="dapr.Dapr",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="DataLoss",grpc_method="InvokeService",grpc_service="dapr.Dapr",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="DataLoss",grpc_method="PublishEvent",grpc_service="dapr.Dapr",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="DataLoss",grpc_method="SaveState",grpc_service="dapr.Dapr",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="DataLoss",grpc_method="UpdateComponent",grpc_service="daprinternal.DaprInternal",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="DeadlineExceeded",grpc_method="CallActor",grpc_service="daprinternal.DaprInternal",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="DeadlineExceeded",grpc_method="CallLocal",grpc_service="daprinternal.DaprInternal",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="DeadlineExceeded",grpc_method="DeleteState",grpc_service="dapr.Dapr",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="DeadlineExceeded",grpc_method="GetState",grpc_service="dapr.Dapr",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="DeadlineExceeded",grpc_method="InvokeBinding",grpc_service="dapr.Dapr",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="DeadlineExceeded",grpc_method="InvokeService",grpc_service="dapr.Dapr",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="DeadlineExceeded",grpc_method="PublishEvent",grpc_service="dapr.Dapr",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="DeadlineExceeded",grpc_method="SaveState",grpc_service="dapr.Dapr",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="DeadlineExceeded",grpc_method="UpdateComponent",grpc_service="daprinternal.DaprInternal",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="FailedPrecondition",grpc_method="CallActor",grpc_service="daprinternal.DaprInternal",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="FailedPrecondition",grpc_method="CallLocal",grpc_service="daprinternal.DaprInternal",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="FailedPrecondition",grpc_method="DeleteState",grpc_service="dapr.Dapr",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="FailedPrecondition",grpc_method="GetState",grpc_service="dapr.Dapr",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="FailedPrecondition",grpc_method="InvokeBinding",grpc_service="dapr.Dapr",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="FailedPrecondition",grpc_method="InvokeService",grpc_service="dapr.Dapr",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="FailedPrecondition",grpc_method="PublishEvent",grpc_service="dapr.Dapr",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="FailedPrecondition",grpc_method="SaveState",grpc_service="dapr.Dapr",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="FailedPrecondition",grpc_method="UpdateComponent",grpc_service="daprinternal.DaprInternal",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="Internal",grpc_method="CallActor",grpc_service="daprinternal.DaprInternal",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="Internal",grpc_method="CallLocal",grpc_service="daprinternal.DaprInternal",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="Internal",grpc_method="DeleteState",grpc_service="dapr.Dapr",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="Internal",grpc_method="GetState",grpc_service="dapr.Dapr",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="Internal",grpc_method="InvokeBinding",grpc_service="dapr.Dapr",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="Internal",grpc_method="InvokeService",grpc_service="dapr.Dapr",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="Internal",grpc_method="PublishEvent",grpc_service="dapr.Dapr",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="Internal",grpc_method="SaveState",grpc_service="dapr.Dapr",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="Internal",grpc_method="UpdateComponent",grpc_service="daprinternal.DaprInternal",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="InvalidArgument",grpc_method="CallActor",grpc_service="daprinternal.DaprInternal",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="InvalidArgument",grpc_method="CallLocal",grpc_service="daprinternal.DaprInternal",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="InvalidArgument",grpc_method="DeleteState",grpc_service="dapr.Dapr",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="InvalidArgument",grpc_method="GetState",grpc_service="dapr.Dapr",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="InvalidArgument",grpc_method="InvokeBinding",grpc_service="dapr.Dapr",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="InvalidArgument",grpc_method="InvokeService",grpc_service="dapr.Dapr",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="InvalidArgument",grpc_method="PublishEvent",grpc_service="dapr.Dapr",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="InvalidArgument",grpc_method="SaveState",grpc_service="dapr.Dapr",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="InvalidArgument",grpc_method="UpdateComponent",grpc_service="daprinternal.DaprInternal",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="NotFound",grpc_method="CallActor",grpc_service="daprinternal.DaprInternal",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="NotFound",grpc_method="CallLocal",grpc_service="daprinternal.DaprInternal",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="NotFound",grpc_method="DeleteState",grpc_service="dapr.Dapr",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="NotFound",grpc_method="GetState",grpc_service="dapr.Dapr",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="NotFound",grpc_method="InvokeBinding",grpc_service="dapr.Dapr",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="NotFound",grpc_method="InvokeService",grpc_service="dapr.Dapr",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="NotFound",grpc_method="PublishEvent",grpc_service="dapr.Dapr",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="NotFound",grpc_method="SaveState",grpc_service="dapr.Dapr",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="NotFound",grpc_method="UpdateComponent",grpc_service="daprinternal.DaprInternal",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="OK",grpc_method="CallActor",grpc_service="daprinternal.DaprInternal",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="OK",grpc_method="CallLocal",grpc_service="daprinternal.DaprInternal",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="OK",grpc_method="DeleteState",grpc_service="dapr.Dapr",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="OK",grpc_method="GetState",grpc_service="dapr.Dapr",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="OK",grpc_method="InvokeBinding",grpc_service="dapr.Dapr",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="OK",grpc_method="InvokeService",grpc_service="dapr.Dapr",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="OK",grpc_method="PublishEvent",grpc_service="dapr.Dapr",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="OK",grpc_method="SaveState",grpc_service="dapr.Dapr",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="OK",grpc_method="UpdateComponent",grpc_service="daprinternal.DaprInternal",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="OutOfRange",grpc_method="CallActor",grpc_service="daprinternal.DaprInternal",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="OutOfRange",grpc_method="CallLocal",grpc_service="daprinternal.DaprInternal",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="OutOfRange",grpc_method="DeleteState",grpc_service="dapr.Dapr",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="OutOfRange",grpc_method="GetState",grpc_service="dapr.Dapr",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="OutOfRange",grpc_method="InvokeBinding",grpc_service="dapr.Dapr",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="OutOfRange",grpc_method="InvokeService",grpc_service="dapr.Dapr",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="OutOfRange",grpc_method="PublishEvent",grpc_service="dapr.Dapr",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="OutOfRange",grpc_method="SaveState",grpc_service="dapr.Dapr",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="OutOfRange",grpc_method="UpdateComponent",grpc_service="daprinternal.DaprInternal",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="PermissionDenied",grpc_method="CallActor",grpc_service="daprinternal.DaprInternal",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="PermissionDenied",grpc_method="CallLocal",grpc_service="daprinternal.DaprInternal",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="PermissionDenied",grpc_method="DeleteState",grpc_service="dapr.Dapr",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="PermissionDenied",grpc_method="GetState",grpc_service="dapr.Dapr",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="PermissionDenied",grpc_method="InvokeBinding",grpc_service="dapr.Dapr",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="PermissionDenied",grpc_method="InvokeService",grpc_service="dapr.Dapr",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="PermissionDenied",grpc_method="PublishEvent",grpc_service="dapr.Dapr",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="PermissionDenied",grpc_method="SaveState",grpc_service="dapr.Dapr",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="PermissionDenied",grpc_method="UpdateComponent",grpc_service="daprinternal.DaprInternal",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="ResourceExhausted",grpc_method="CallActor",grpc_service="daprinternal.DaprInternal",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="ResourceExhausted",grpc_method="CallLocal",grpc_service="daprinternal.DaprInternal",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="ResourceExhausted",grpc_method="DeleteState",grpc_service="dapr.Dapr",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="ResourceExhausted",grpc_method="GetState",grpc_service="dapr.Dapr",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="ResourceExhausted",grpc_method="InvokeBinding",grpc_service="dapr.Dapr",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="ResourceExhausted",grpc_method="InvokeService",grpc_service="dapr.Dapr",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="ResourceExhausted",grpc_method="PublishEvent",grpc_service="dapr.Dapr",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="ResourceExhausted",grpc_method="SaveState",grpc_service="dapr.Dapr",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="ResourceExhausted",grpc_method="UpdateComponent",grpc_service="daprinternal.DaprInternal",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="Unauthenticated",grpc_method="CallActor",grpc_service="daprinternal.DaprInternal",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="Unauthenticated",grpc_method="CallLocal",grpc_service="daprinternal.DaprInternal",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="Unauthenticated",grpc_method="DeleteState",grpc_service="dapr.Dapr",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="Unauthenticated",grpc_method="GetState",grpc_service="dapr.Dapr",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="Unauthenticated",grpc_method="InvokeBinding",grpc_service="dapr.Dapr",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="Unauthenticated",grpc_method="InvokeService",grpc_service="dapr.Dapr",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="Unauthenticated",grpc_method="PublishEvent",grpc_service="dapr.Dapr",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="Unauthenticated",grpc_method="SaveState",grpc_service="dapr.Dapr",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="Unauthenticated",grpc_method="UpdateComponent",grpc_service="daprinternal.DaprInternal",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="Unavailable",grpc_method="CallActor",grpc_service="daprinternal.DaprInternal",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="Unavailable",grpc_method="CallLocal",grpc_service="daprinternal.DaprInternal",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="Unavailable",grpc_method="DeleteState",grpc_service="dapr.Dapr",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="Unavailable",grpc_method="GetState",grpc_service="dapr.Dapr",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="Unavailable",grpc_method="InvokeBinding",grpc_service="dapr.Dapr",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="Unavailable",grpc_method="InvokeService",grpc_service="dapr.Dapr",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="Unavailable",grpc_method="PublishEvent",grpc_service="dapr.Dapr",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="Unavailable",grpc_method="SaveState",grpc_service="dapr.Dapr",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="Unavailable",grpc_method="UpdateComponent",grpc_service="daprinternal.DaprInternal",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="Unimplemented",grpc_method="CallActor",grpc_service="daprinternal.DaprInternal",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="Unimplemented",grpc_method="CallLocal",grpc_service="daprinternal.DaprInternal",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="Unimplemented",grpc_method="DeleteState",grpc_service="dapr.Dapr",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="Unimplemented",grpc_method="GetState",grpc_service="dapr.Dapr",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="Unimplemented",grpc_method="InvokeBinding",grpc_service="dapr.Dapr",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="Unimplemented",grpc_method="InvokeService",grpc_service="dapr.Dapr",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="Unimplemented",grpc_method="PublishEvent",grpc_service="dapr.Dapr",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="Unimplemented",grpc_method="SaveState",grpc_service="dapr.Dapr",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="Unimplemented",grpc_method="UpdateComponent",grpc_service="daprinternal.DaprInternal",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="Unknown",grpc_method="CallActor",grpc_service="daprinternal.DaprInternal",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="Unknown",grpc_method="CallLocal",grpc_service="daprinternal.DaprInternal",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="Unknown",grpc_method="DeleteState",grpc_service="dapr.Dapr",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="Unknown",grpc_method="GetState",grpc_service="dapr.Dapr",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="Unknown",grpc_method="InvokeBinding",grpc_service="dapr.Dapr",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="Unknown",grpc_method="InvokeService",grpc_service="dapr.Dapr",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="Unknown",grpc_method="PublishEvent",grpc_service="dapr.Dapr",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="Unknown",grpc_method="SaveState",grpc_service="dapr.Dapr",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="Unknown",grpc_method="UpdateComponent",grpc_service="daprinternal.DaprInternal",grpc_type="unary"} 0
# HELP grpc_server_msg_received_total Total number of RPC stream messages received on the server.
# TYPE grpc_server_msg_received_total counter
grpc_server_msg_received_total{grpc_method="CallActor",grpc_service="daprinternal.DaprInternal",grpc_type="unary"} 0
grpc_server_msg_received_total{grpc_method="CallLocal",grpc_service="daprinternal.DaprInternal",grpc_type="unary"} 0
grpc_server_msg_received_total{grpc_method="DeleteState",grpc_service="dapr.Dapr",grpc_type="unary"} 0
grpc_server_msg_received_total{grpc_method="GetState",grpc_service="dapr.Dapr",grpc_type="unary"} 0
grpc_server_msg_received_total{grpc_method="InvokeBinding",grpc_service="dapr.Dapr",grpc_type="unary"} 0
grpc_server_msg_received_total{grpc_method="InvokeService",grpc_service="dapr.Dapr",grpc_type="unary"} 0
grpc_server_msg_received_total{grpc_method="PublishEvent",grpc_service="dapr.Dapr",grpc_type="unary"} 0
grpc_server_msg_received_total{grpc_method="SaveState",grpc_service="dapr.Dapr",grpc_type="unary"} 0
grpc_server_msg_received_total{grpc_method="UpdateComponent",grpc_service="daprinternal.DaprInternal",grpc_type="unary"} 0
# HELP grpc_server_msg_sent_total Total number of gRPC stream messages sent by the server.
# TYPE grpc_server_msg_sent_total counter
grpc_server_msg_sent_total{grpc_method="CallActor",grpc_service="daprinternal.DaprInternal",grpc_type="unary"} 0
grpc_server_msg_sent_total{grpc_method="CallLocal",grpc_service="daprinternal.DaprInternal",grpc_type="unary"} 0
grpc_server_msg_sent_total{grpc_method="DeleteState",grpc_service="dapr.Dapr",grpc_type="unary"} 0
grpc_server_msg_sent_total{grpc_method="GetState",grpc_service="dapr.Dapr",grpc_type="unary"} 0
grpc_server_msg_sent_total{grpc_method="InvokeBinding",grpc_service="dapr.Dapr",grpc_type="unary"} 0
grpc_server_msg_sent_total{grpc_method="InvokeService",grpc_service="dapr.Dapr",grpc_type="unary"} 0
grpc_server_msg_sent_total{grpc_method="PublishEvent",grpc_service="dapr.Dapr",grpc_type="unary"} 0
grpc_server_msg_sent_total{grpc_method="SaveState",grpc_service="dapr.Dapr",grpc_type="unary"} 0
grpc_server_msg_sent_total{grpc_method="UpdateComponent",grpc_service="daprinternal.DaprInternal",grpc_type="unary"} 0
# HELP grpc_server_started_total Total number of RPCs started on the server.
# TYPE grpc_server_started_total counter
grpc_server_started_total{grpc_method="CallActor",grpc_service="daprinternal.DaprInternal",grpc_type="unary"} 0
grpc_server_started_total{grpc_method="CallLocal",grpc_service="daprinternal.DaprInternal",grpc_type="unary"} 0
grpc_server_started_total{grpc_method="DeleteState",grpc_service="dapr.Dapr",grpc_type="unary"} 0
grpc_server_started_total{grpc_method="GetState",grpc_service="dapr.Dapr",grpc_type="unary"} 0
grpc_server_started_total{grpc_method="InvokeBinding",grpc_service="dapr.Dapr",grpc_type="unary"} 0
grpc_server_started_total{grpc_method="InvokeService",grpc_service="dapr.Dapr",grpc_type="unary"} 0
grpc_server_started_total{grpc_method="PublishEvent",grpc_service="dapr.Dapr",grpc_type="unary"} 0
grpc_server_started_total{grpc_method="SaveState",grpc_service="dapr.Dapr",grpc_type="unary"} 0
grpc_server_started_total{grpc_method="UpdateComponent",grpc_service="daprinternal.DaprInternal",grpc_type="unary"} 0
# HELP http_handler_completed_latency_seconds Latency of completed requests.
# TYPE http_handler_completed_latency_seconds histogram
http_handler_completed_latency_seconds_bucket{handler="",host="localhost:3500",method="POST",name="jjcollinge-daprd",path="/v1.0/state/statestore",status="200",le="0.01"} 5
http_handler_completed_latency_seconds_bucket{handler="",host="localhost:3500",method="POST",name="jjcollinge-daprd",path="/v1.0/state/statestore",status="200",le="0.03"} 5
http_handler_completed_latency_seconds_bucket{handler="",host="localhost:3500",method="POST",name="jjcollinge-daprd",path="/v1.0/state/statestore",status="200",le="0.1"} 5
http_handler_completed_latency_seconds_bucket{handler="",host="localhost:3500",method="POST",name="jjcollinge-daprd",path="/v1.0/state/statestore",status="200",le="0.3"} 5
http_handler_completed_latency_seconds_bucket{handler="",host="localhost:3500",method="POST",name="jjcollinge-daprd",path="/v1.0/state/statestore",status="200",le="1"} 5
http_handler_completed_latency_seconds_bucket{handler="",host="localhost:3500",method="POST",name="jjcollinge-daprd",path="/v1.0/state/statestore",status="200",le="3"} 5
http_handler_completed_latency_seconds_bucket{handler="",host="localhost:3500",method="POST",name="jjcollinge-daprd",path="/v1.0/state/statestore",status="200",le="10"} 5
http_handler_completed_latency_seconds_bucket{handler="",host="localhost:3500",method="POST",name="jjcollinge-daprd",path="/v1.0/state/statestore",status="200",le="30"} 5
http_handler_completed_latency_seconds_bucket{handler="",host="localhost:3500",method="POST",name="jjcollinge-daprd",path="/v1.0/state/statestore",status="200",le="100"} 5
http_handler_completed_latency_seconds_bucket{handler="",host="localhost:3500",method="POST",name="jjcollinge-daprd",path="/v1.0/state/statestore",status="200",le="300"} 5
http_handler_completed_latency_seconds_bucket{handler="",host="localhost:3500",method="POST",name="jjcollinge-daprd",path="/v1.0/state/statestore",status="200",le="+Inf"} 5
http_handler_completed_latency_seconds_sum{handler="",host="localhost:3500",method="POST",name="jjcollinge-daprd",path="/v1.0/state/statestore",status="200"} 0.0100152
http_handler_completed_latency_seconds_count{handler="",host="localhost:3500",method="POST",name="jjcollinge-daprd",path="/v1.0/state/statestore",status="200"} 5
# HELP http_handler_completed_requests_total Count of completed requests.
# TYPE http_handler_completed_requests_total counter
http_handler_completed_requests_total{handler="",host="localhost:3500",method="POST",name="jjcollinge-daprd",path="/v1.0/state/statestore",status="200"} 5
# HELP http_handler_request_size_bytes Size of received requests.
# TYPE http_handler_request_size_bytes histogram
http_handler_request_size_bytes_bucket{handler="",host="localhost:3500",method="POST",name="jjcollinge-daprd",path="/v1.0/state/statestore",le="32"} 0
http_handler_request_size_bytes_bucket{handler="",host="localhost:3500",method="POST",name="jjcollinge-daprd",path="/v1.0/state/statestore",le="1024"} 5
http_handler_request_size_bytes_bucket{handler="",host="localhost:3500",method="POST",name="jjcollinge-daprd",path="/v1.0/state/statestore",le="32768"} 5
http_handler_request_size_bytes_bucket{handler="",host="localhost:3500",method="POST",name="jjcollinge-daprd",path="/v1.0/state/statestore",le="1.048576e+06"} 5
http_handler_request_size_bytes_bucket{handler="",host="localhost:3500",method="POST",name="jjcollinge-daprd",path="/v1.0/state/statestore",le="3.3554432e+07"} 5
http_handler_request_size_bytes_bucket{handler="",host="localhost:3500",method="POST",name="jjcollinge-daprd",path="/v1.0/state/statestore",le="1.073741824e+09"} 5
http_handler_request_size_bytes_bucket{handler="",host="localhost:3500",method="POST",name="jjcollinge-daprd",path="/v1.0/state/statestore",le="+Inf"} 5
http_handler_request_size_bytes_sum{handler="",host="localhost:3500",method="POST",name="jjcollinge-daprd",path="/v1.0/state/statestore"} 195
http_handler_request_size_bytes_count{handler="",host="localhost:3500",method="POST",name="jjcollinge-daprd",path="/v1.0/state/statestore"} 5
# HELP http_handler_response_size_bytes Size of sent responses.
# TYPE http_handler_response_size_bytes histogram
http_handler_response_size_bytes_bucket{handler="",host="localhost:3500",method="POST",name="jjcollinge-daprd",path="/v1.0/state/statestore",status="200",le="32"} 5
http_handler_response_size_bytes_bucket{handler="",host="localhost:3500",method="POST",name="jjcollinge-daprd",path="/v1.0/state/statestore",status="200",le="1024"} 5
http_handler_response_size_bytes_bucket{handler="",host="localhost:3500",method="POST",name="jjcollinge-daprd",path="/v1.0/state/statestore",status="200",le="32768"} 5
http_handler_response_size_bytes_bucket{handler="",host="localhost:3500",method="POST",name="jjcollinge-daprd",path="/v1.0/state/statestore",status="200",le="1.048576e+06"} 5
http_handler_response_size_bytes_bucket{handler="",host="localhost:3500",method="POST",name="jjcollinge-daprd",path="/v1.0/state/statestore",status="200",le="3.3554432e+07"} 5
http_handler_response_size_bytes_bucket{handler="",host="localhost:3500",method="POST",name="jjcollinge-daprd",path="/v1.0/state/statestore",status="200",le="1.073741824e+09"} 5
http_handler_response_size_bytes_bucket{handler="",host="localhost:3500",method="POST",name="jjcollinge-daprd",path="/v1.0/state/statestore",status="200",le="+Inf"} 5
http_handler_response_size_bytes_sum{handler="",host="localhost:3500",method="POST",name="jjcollinge-daprd",path="/v1.0/state/statestore",status="200"} 0
http_handler_response_size_bytes_count{handler="",host="localhost:3500",method="POST",name="jjcollinge-daprd",path="/v1.0/state/statestore",status="200"} 5
# HELP http_handler_started_requests_total Count of started requests.
# TYPE http_handler_started_requests_total counter
http_handler_started_requests_total{handler="",host="localhost:3500",method="POST",name="jjcollinge-daprd",path="/v1.0/state/statestore"} 5
# HELP process_cpu_seconds_total Total user and system CPU time spent in seconds.
# TYPE process_cpu_seconds_total counter
process_cpu_seconds_total 0.56
# HELP process_max_fds Maximum number of open file descriptors.
# TYPE process_max_fds gauge
process_max_fds 4096
# HELP process_open_fds Number of open file descriptors.
# TYPE process_open_fds gauge
process_open_fds 16
# HELP process_resident_memory_bytes Resident memory size in bytes.
# TYPE process_resident_memory_bytes gauge
process_resident_memory_bytes 6.320128e+07
# HELP process_start_time_seconds Start time of the process since unix epoch in seconds.
# TYPE process_start_time_seconds gauge
process_start_time_seconds 1.58124128815e+09
# HELP process_virtual_memory_bytes Virtual memory size in bytes.
# TYPE process_virtual_memory_bytes gauge
process_virtual_memory_bytes 1.574658048e+09
# HELP process_virtual_memory_max_bytes Maximum amount of virtual memory available in bytes.
# TYPE process_virtual_memory_max_bytes gauge
process_virtual_memory_max_bytes -1
# HELP promhttp_metric_handler_requests_in_flight Current number of scrapes being served.
# TYPE promhttp_metric_handler_requests_in_flight gauge
promhttp_metric_handler_requests_in_flight 1
# HELP promhttp_metric_handler_requests_total Total number of scrapes by HTTP status code.
# TYPE promhttp_metric_handler_requests_total counter
promhttp_metric_handler_requests_total{code="200"} 5
promhttp_metric_handler_requests_total{code="500"} 0
promhttp_metric_handler_requests_total{code="503"} 0
```

youngbupark commented:
@jjcollinge Thanks for all your change. 

There might be miscommunication:) I think adding metricspec is out of scope of this PR. 

> Also let's continue to discuss MetricSpec in #972. We should have a mini-spec and review it with the other contributors.

As I mentioned above, this needs to be spec'ed properly and reviewed with the other maintainers in https://github.com/dapr/dapr/issues/972 because we need to standardize all three telemetry specs (traces, logs, metrics) and to generate k8s clients for dapr for CRD.

So, please exclude the changes associated with #972 (Grouping and CRD change). For now, I would focus on exposing metrics endpoint with minimal metrics, making prometheus endpoint optional via a command argument like you did in the beginning, and adding more **unit-tests**.  When it comes to grouping concept and crd change, you can add them iteratively as a part of #972 .

Please keep PR small and add unit-tests or e2e test relatively.

youngbupark commented:
> Note to self, I’m not sure this is the right approach and will explore moving to using a dedicated server for the metrics endpoint.

It makes sense to me, but we need to make it optional via command line as you did in the first change. From my research,  I found that some customers' scenarios use push mode without prometheus.

@yaron2 @Haishi2016 Any objections to creating new http server dedicated to prometheus endpoint?

cc/ @amanbha @artursouza

yaron2 commented:
I don't have any problem with a dedicated server, as it's a common practice.

If we start it via a flag, we also need to decide if we enable the metrics server by default, which I think we should.

I support focusing this PR on exposing the metrics server via flags. It'll be a big win and allow us to move forward iteratively.

yaron2 commented:
I think we can use the MetricSpec proposed here as the basis for #972.

Would this mean that this PR will assume as if `UseDefaults` was set to `true`?

jjcollinge commented:
@youngbupark thanks for clarifying! Happy to back out the metrics spec in favour of an `enableMetrics` and `metricsPort` flag with defaults `true` and `9090` respectively. Sould I keep both the current gRPC and HTTP metrics middleware in?

yaron2 commented:
> Sould I keep both the current gRPC and HTTP metrics middleware in?

Yes. Do you see a reason not to?

youngbupark commented:
> @youngbupark thanks for clarifying! Happy to back out the metrics spec in favour of an `enableMetrics` and `metricsPort` flag with defaults `true` and `9090` respectively. Sould I keep both the current gRPC and HTTP metrics middleware in?

Yes, we should. I definitely appreciated your change adding grpc and http metrics middleware! :) The only part I would discuss more is MetricSpec CRD. Also I always encourage anyone to create small PRs and add proper unit-tests:
1) small PR allows to review the pr easier for anyone who are unfamiliar with the code and reduce the mistake for author.
2) squash merge destroys git commit history
3) unit-test is crucial for production ready product.

for these reasons, I recommend you to keep your metricspec change and create the separate PR :)

youngbupark commented on pkg/diagnostics/metrics_test.go:
```diff
@@ -0,0 +1,130 @@
+		expectedMap[metric] = false
+	}
+
+	// Remove metrics not on platform
```
nit: Please remove this comment.

youngbupark commented:
@jjcollinge Now we can merge your pr, but it is blocked because of conflicting in go.mod. Please resolve it :) Then we will merge it! Thanks!
